### PR TITLE
[RTM] Move the user session handling to a separate listener

### PR DESCRIPTION
**Pull request by Toflar**
Follow-Up PR to https://github.com/contao/core-bundle/pull/184.

Now also contains the FrontendUser handling and does the loading process as well. In addition, `TL_REFERER_ID` has been replaced as well.

Tests still to be added.

### CODE REVIEW

**Comment by ausi** [comment]
This PR fixes #206. But the session doesn’t work for the theme import in [Theme.php:78](https://github.com/Toflar/core-bundle/blob/a741505a7a0602e464dc98bfe64b82586e6b1afa/src/Resources/contao/classes/Theme.php#L78) and [Theme.php:107](https://github.com/Toflar/core-bundle/blob/a741505a7a0602e464dc98bfe64b82586e6b1afa/src/Resources/contao/classes/Theme.php#L107). Is this issue related to this PR or should I create a new ticket for it?

**Comment by Toflar** [comment]
The legacy `Session` class is transparently mapped. Why does it not work? See https://github.com/contao/core-bundle/pull/68/files#diff-2f0eb5d18e0f5fb0dbaef526bb165d41R415

**Comment by ausi** [comment]
It does work if I remove the line [UserSessionListener.php:163](https://github.com/Toflar/core-bundle/blob/a741505a7a0602e464dc98bfe64b82586e6b1afa/src/EventListener/UserSessionListener.php#L163) `!$request->query->has('key')`. The URL of the theme import is `/contao?do=themes&key=importTheme&rt=...&ref=...`

**Comment by Toflar** [comment]
And why does it work in 3.5? https://github.com/contao/core/blob/master/system/modules/core/classes/BackendUser.php#L108

**Comment by ausi** [comment]
Because `$this->Database->prepare("UPDATE...` is outside of the `if` in Contao 3.5: https://github.com/contao/core/blob/8f678aaf08102432f943241a90d13c1c308741bd/system/modules/core/classes/BackendUser.php#L150-L155

**Comment by Toflar** [comment]
Okay, so storing the session should always be done.

**Comment by Toflar** [comment]
Can you try c2cd469?

**Comment by ausi** [comment]
Theme import works with c2cd469797386f10b02f29fb4648ec18071f3d32, thanks!

But it seems there are problems with the referrer management. If you edit a content element and hit “save & close” it takes you back to the backend home page.

**Comment by bytehead** [comment]
See #143

**Comment by Toflar** [comment]
Should work in c04d6fe.

**Comment by ausi** [comment]
> Should work in c04d6fe96a7099485728fa0b33f3d34c2869896d.

Referrer management works now, thanks again!

File: src/EventListener/UserSessionListener.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+    /**
+     * Checks whether the user has session data stored and passes it to the
+     * session service if so.
+     *
+     * @param GetResponseEvent $event The event object
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        if (!$this->isFrontendMasterRequest($event)
```
Comment (aschempp) [inline src/EventListener/UserSessionListener.php]: Will that mean we do not have a session on subrequests?

File: src/EventListener/UserSessionListener.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    /**
+     * Stores the session for the front end.
+     *
+     * @param Request $request
+     */
+    private function storeFrontendSession(Request $request)
+    {
+        /** @var AttributeBagInterface $bag */
```
Comment (aschempp) [inline src/EventListener/UserSessionListener.php]: Why are these type hints necessary if the method has a correct `@return` block?

File: src/EventListener/UserSessionListener.php [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        return $this->session->getBag($bag);
+    }
+
+    /**
+     * Stores the session data in the database.
+     */
+    private function storeSession()
+    {
+        $user   = $this->getUserObject();
+        $table  = $this->isFrontendScope() ? 'tl_member' : 'tl_user';
```
Comment (aschempp) [inline src/EventListener/UserSessionListener.php]: It would be better to add a `getTable` method to the user object.

Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: I don't want to change the legacy code too much.

File: src/EventListener/UserSessionListener.php (hunk 2 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: Ah, that was needed before I put this in its own method. Removed in 4bb2e5c952f443529c76332796389cb12ccaf011.

File: src/EventListener/UserSessionListener.php (hunk 1 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: Yes, which is correct in my opinion. If you want to have session data you should take the session of the master request.

File: src/Resources/config/listener.yml [hunk 4]
```diff
@@ -87,3 +87,17 @@ services:
...
+        calls:
+            - [setContainer, ["@service_container"]]
+        arguments:
+            - '@session'
+            - '@doctrine.dbal.default_connection'
+        tags:
+            # Priority must be lower than contao.listener.session (64)
+            - { name: kernel.event_listener, event: kernel.request, method: onKernelRequest }
+            # Priority must be higher than the framework bundle save session listener (-1000)
+            - { name: kernel.event_listener, event: kernel.response, method: onKernelResponse }
```
Comment (leofeyer) [inline src/Resources/config/listener.yml]: I assume that Symfony uses the `kernel.response` event as well and you have copied it from there? Otherwise, why don't we use the `kernel.terminate` event?

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (leofeyer) [inline src/EventListener/UserSessionListener.php]: Since `User::storeSession()` is a public function, why don't you just do `$this->getUserObject()->storeSession()`? We could leave most of the legacy code in the `*User` classes then.

File: src/Resources/contao/library/Contao/Controller.php [hunk 5]
```diff
@@ -1003,6 +1003,7 @@ public static function redirect($strLocation, $intStatus=303)
 			exit;
 		}
 
+		$strLocation = ltrim($strLocation, '/');
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/Controller.php]: Why is this necessary?

File: src/Resources/config/listener.yml (hunk 4 above)
Comment (Toflar) [inline src/Resources/config/listener.yml]: I did not copy, no. Actually, this is the whole thing about the problem. We have to read and store away the session into the database before the framework bundle listener closes the session.

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: Because I think we want to keep this functionality even if we remove our user classes once. That way, the work is already done.

File: src/Resources/config/listener.yml (hunk 4 above)
Comment (leofeyer) [inline src/Resources/config/listener.yml]: And the framework bundle listener listens to the `kernel.response` event?

Comment (Toflar) [inline src/Resources/config/listener.yml]: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/HttpKernel/EventListener/SaveSessionListener.php#L63

File: src/Resources/contao/library/Contao/Controller.php (hunk 5 above)
Comment (Toflar) [inline src/Resources/contao/library/Contao/Controller.php]: Because `$request->getRequestUri()` returns the URI including the leading `/` which is correct. Our `Controller::redirect()` method would turn things into something like that: `http://domain.de//path/to/whatever`. I thought it's better to ensure there's never a leading `/` than doing stupid `str_replace()` calls **everytime** we pass something into that legacy method.

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (leofeyer) [inline src/EventListener/UserSessionListener.php]: Not sure about that. Look at this issue in our [Ahrweiler-TODO](https://github.com/contao/symfony-todo/issues/3):

> Nur State-Informationen in die DB-Session, Referer und Token in `$_SESSION`

This means that we have to rework the session handling, therefore I don't think we will keep the current code.

We should either keep this change as simple as possible or incorporate the new feature, shouldn't we?

File: src/Resources/config/listener.yml (hunk 4 above)
Comment (leofeyer) [inline src/Resources/config/listener.yml]: Ok, then this is fine :)

File: src/Resources/contao/library/Contao/Controller.php (hunk 5 above)
Comment (leofeyer) [inline src/Resources/contao/library/Contao/Controller.php]: But this also means that we won't be able to detect wrong input anymore. Also, I think the `ltrim()` call should be on line 1013:

```php
// Make the location an absolute URL
if (!preg_match('@^https?://@i', $strLocation))
{
	$strLocation = \Environment::get('base') . ltrim($strLocation, '/');
}
```

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: Well, basically this is already done then :-) If you look at the code carefully you notice that this works well together with my first PR that integrated the session using the session bags `contao_frontend` and `contao_backend`. The listener here **only** considers those two!! This results in the easy fact that if you want to have Contao store away your session values you add it to either the `contao_frontend` or the `contao_backend` AttributeBag. If you want to have it as stateless you just write to the default SF2 attribute bags or your own :-)

File: src/Resources/contao/library/Contao/Controller.php (hunk 5 above)
Comment (Toflar) [inline src/Resources/contao/library/Contao/Controller.php]: Why would you want to detect wrong input. The method just doesn't work if you pass a leading `/` so why do you think it's wrong if it ensures that there's none automatically? And yes, we could move it down there if you prefer.

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (leofeyer) [inline src/EventListener/UserSessionListener.php]: Ok, that's great. But I'd still prefer to leave the legacy code in the `*User` classes. It would also make it easier to synchronize it with the Contao 3 repository.

File: src/Resources/contao/library/Contao/Controller.php (hunk 5 above)
Comment (leofeyer) [inline src/Resources/contao/library/Contao/Controller.php]: > Why would you want to detect wrong input

Obviously, because I would want to fix it :)

Comment (Toflar) [inline src/Resources/contao/library/Contao/Controller.php]: That only makes sense if the automated fixing could generate false positives. But in this case it cannot. 
Your proposal should look like this

```php
if ($strLocation[0] == '/')
{
    throw new InvalidArgumentException('The target location string contains a leading slash which it must not!');
}
```

**THAT** would be detecting wrong input. Right now it just does nothing. It passes it on and generates wrong URL's. Either you fix it automatically (which you can do in this case) or you have to throw an Exception :-)

Comment (leofeyer) [inline src/Resources/contao/library/Contao/Controller.php]: You are right. Then keep it as is and only move it down to line 1013.

Comment (Toflar) [inline src/Resources/contao/library/Contao/Controller.php]: Okay. Done in ea0299a156f67d5e79da6698a1c91c2ce7168f57.

File: src/EventListener/UserSessionListener.php [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+    /**
+     * Stores the session for the back end.
+     *
+     * @param Request $request
+     */
+    private function storeBackendSession(Request $request)
+    {
+        if (!$this->canModifyBackendSession($request)) {
+            $this->storeSession();
```
Comment (leofeyer) [inline src/EventListener/UserSessionListener.php]: Why are you calling `$this->storeSession()` if the session cannot be modified anyway?

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (aschempp) [inline src/EventListener/UserSessionListener.php]: I'm for keeping @Toflar's code, because we will need the new code at some point (e.g. when introducing advanced security).

Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: The old code uses old classes such as `Database`, `Session` and `Input` and the constant  `TL_REFERER_ID`. I already did all the refactoring work and it's a lot better now and we can even write tests for it. Especially as synchronizing with the Contao 3 repository is not an argument at all because if it was one it would mean that we can never move on to new things because if we do syncing with Contao 3 is complicated :-)

File: src/EventListener/UserSessionListener.php (hunk 6 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: Because you always need to store the data of the session? We only modify the referer part but there are a lot of other things going on. This is also done in Contao 3 (storing is outside the `if` clause). Also see @ausi's comment here: https://github.com/contao/core-bundle/pull/207#issuecomment-93442689

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (leofeyer) [inline src/EventListener/UserSessionListener.php]: Ok, but then you might as well add the `getTable()` method as @aschempp mentioned above :)

**Comment by leofeyer** [comment]
Looks good to me. @Toflar Do you happen to have time to move the referer handling to a separate session bag, which is not stored in the database? The same applies to the request token and the referer ID.

Or would you prefer to create a separate PR for this?

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: That's true, shall I?

**Comment by Toflar** [comment]
> Looks good to me. @Toflar Do you happen to have time to move the referer handling to a separate session bag, which is not stored in the database? The same applies to the request token and the referer ID.

I wanted to do some more work on the weekend but would have spent that on writing tests for this PR because I think it's more important to finish this first. What's your opinion?

**Comment by Toflar** [comment]
> Or would you prefer to create a separate PR for this?

Oh, and yes!

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (leofeyer) [inline src/EventListener/UserSessionListener.php]: +1

**Comment by leofeyer** [comment]
> I wanted to do some more work on the weekend but would have spent that on writing tests for this PR because I think it's more important to finish this first. What's your opinion?

+1

File: src/EventListener/UserSessionListener.php (hunk 3 above)
Comment (Toflar) [inline src/EventListener/UserSessionListener.php]: Done in ec4b90a3c68c3e16463c5d049ec9562913304c0e.

File: src/Resources/contao/library/Contao/Controller.php (hunk 5 above)
Comment (aschempp) [inline src/Resources/contao/library/Contao/Controller.php]: Sorry, but this must be wrong. If the input contains a leading slash, it's an absolute URL (probably generated by the Symfony router). Removing the leading slash does not (always) correctly convert an absolute to a relative URL! We had exactly the same problems with TL_SCRIPT.

Comment (Toflar) [inline src/Resources/contao/library/Contao/Controller.php]: Fixed in ed1a2b37d8f5cb7fa84b1c58d9764e401b7cfc43.

**Comment by leofeyer** [comment]
Is this RTM then?

**Comment by Toflar** [comment]
Like 3 hours ago you were +1 on writing tests for it? How can it possibly be RTM now??

**Comment by leofeyer** [comment]
I though you wanted to add the tests on the weekend, rather than implementing the new feature. But I was not aware that I have to wait for the tests to be completed to merge this PR.

But it probably makes sense this way :smiley:

**Comment by Toflar** [comment]
@discordier, @leofeyer: I tried to mock `__get()` for 2.5 hours now and I did it in any possible way but I just can't figure out why it doesn't work :-( Giving up for today.

**Comment by discordier** [comment]
@Toflar see https://github.com/Toflar/core-bundle/pull/1 ;)

When mocking special methods like `__get()` which have a coded arguments, you have to provide a base class implementing it. This was missing in our fixture class.

I also fixed a second problem which happened to occur in `Contao\CoreBundle\Test\Security\Authentication\ContaoAuthenticatorTest::testAuthenticateToken()` and tried to use `Contao\User` as `User` while the name was already been taken.
This is most likely still due our massive usage of `class_alias()`.

The error was: `PHP Fatal error:  Cannot use Contao\User as User because the name is already in use in tests/Fixtures/library/FrontendUser.php on line 5`

**Comment by Toflar** [comment]
Ah again because of those stupid fixtures. Can't we get rid of them finally? @aschempp: Any reason why you are using fixtures in your tests instead of mocking the classes?

**Comment by leofeyer** [comment]
They are actually dummy classes and not fixtures. And I don't think we can get rid of them easily. Couldn't write tests for the event listener without using dummy classes :(

**Comment by Toflar** [comment]
PHPUnit can mock classes that don't exist using `getMockBuilder()`. We **must** get rid of those dummy classes using the mock builder and/or adapter classes. Green PHPUnit outputs won't be satisfying me as long as we have unpredictable dummy classes in our test suite. But that's not something you need to care about. I've added an issue https://github.com/contao/core-bundle/issues/226 so I don't forget about it.

**Comment by Toflar** [comment]
\*phew* - this is as good as I can get it for now. It's 94.39% coverage on lines. We can still add tests but I guess it's good enough to merge now. Changing to RTM.
